File: src/csharp/Intel/Iced/Intel/MvexInfo.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if (DECODER || ENCODER || (ENCODER && OPCODE_INFO)) && MVEX
using System.Diagnostics;

namespace Iced.Intel {
	readonly struct MvexInfo {
		readonly int index;

		public MvexTupleTypeLutKind TupleTypeLutKind => (MvexTupleTypeLutKind)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.TupleTypeLutKindIndex];
		public MvexEHBit EHBit => (MvexEHBit)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.EHBitIndex];
		public MvexConvFn ConvFn => (MvexConvFn)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.ConvFnIndex];
		public uint InvalidConvFns => MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.InvalidConvFnsIndex];
		public uint InvalidSwizzleFns => MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.InvalidSwizzleFnsIndex];
		public bool IsNDD => ((MvexInfoFlags1)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.Flags1Index] & MvexInfoFlags1.NDD) != 0;
		public bool IsNDS => ((MvexInfoFlags1)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.Flags1Index] & MvexInfoFlags1.NDS) != 0;
		public bool CanUseEvictionHint => ((MvexInfoFlags1)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.Flags1Index] & MvexInfoFlags1.EvictionHint) != 0;
		public bool CanUseImmRoundingControl => ((MvexInfoFlags1)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.Flags1Index] & MvexInfoFlags1.ImmRoundingControl) != 0;
		public bool CanUseRoundingControl => ((MvexInfoFlags1)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.Flags1Index] & MvexInfoFlags1.RoundingControl) != 0;
		public bool CanUseSuppressAllExceptions => ((MvexInfoFlags1)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.Flags1Index] & MvexInfoFlags1.SuppressAllExceptions) != 0;
		public bool IgnoresOpMaskRegister => ((MvexInfoFlags1)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.Flags1Index] & MvexInfoFlags1.IgnoresOpMaskRegister) != 0;
		public bool RequireOpMaskRegister => ((MvexInfoFlags1)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.Flags1Index] & MvexInfoFlags1.RequireOpMaskRegister) != 0;
		public bool NoSaeRc => ((MvexInfoFlags2)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.Flags2Index] & MvexInfoFlags2.NoSaeRoundingControl) != 0;
		public bool IsConvFn32 => ((MvexInfoFlags2)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.Flags2Index] & MvexInfoFlags2.ConvFn32) != 0;
		public bool IgnoresEvictionHint => ((MvexInfoFlags2)MvexInfoData.Data[index * MvexInfoData.StructSize + MvexInfoData.Flags2Index] & MvexInfoFlags2.IgnoresEvictionHint) != 0;

		public MvexInfo(Code code) {
			index = (int)code - (int)IcedConstants.MvexStart;
			Debug.Assert((uint)index < IcedConstants.MvexLength);
		}
		
		public TupleType GetTupleType(int sss) => (TupleType)MvexTupleTypeLut.Data[(int)TupleTypeLutKind * 8 + sss];
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/internal/MvexInfo.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.internal;

/** DO NOT USE: INTERNAL API */
public final class MvexInfo {
	private static final byte[] data = ResourceReader.readByteArray(MvexInfo.class.getClassLoader(),
			"com/github/icedland/iced/x86/MvexInfoData.bin");

	/** DO NOT USE: INTERNAL API */
	public static boolean isMvex(int code) {
		return Integer.compareUnsigned(code - IcedConstants.MVEX_START, IcedConstants.MVEX_LENGTH) < 0;
	}

	/** DO NOT USE: INTERNAL API */
	public static int getTupleTypeLutKind(int code) {
		int index = code - IcedConstants.MVEX_START;
		return data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.TUPLE_TYPE_LUT_KIND_INDEX];
	}

	/** DO NOT USE: INTERNAL API */
	public static int getEHBit(int code) {
		int index = code - IcedConstants.MVEX_START;
		return data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.EH_BIT_INDEX];
	}

	/** DO NOT USE: INTERNAL API */
	public static int getConvFn(int code) {
		int index = code - IcedConstants.MVEX_START;
		return data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.CONV_FN_INDEX];
	}

	/** DO NOT USE: INTERNAL API */
	public static int getInvalidConvFns(int code) {
		int index = code - IcedConstants.MVEX_START;
		return data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.INVALID_CONV_FNS_INDEX];
	}

	/** DO NOT USE: INTERNAL API */
	public static int getInvalidSwizzleFns(int code) {
		int index = code - IcedConstants.MVEX_START;
		return data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.INVALID_SWIZZLE_FNS_INDEX];
	}

	/** DO NOT USE: INTERNAL API */
	public static boolean isNDD(int code) {
		int index = code - IcedConstants.MVEX_START;
		return (data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.FLAGS1_INDEX] & MvexInfoFlags1.NDD) != 0;
	}

	/** DO NOT USE: INTERNAL API */
	public static boolean isNDS(int code) {
		int index = code - IcedConstants.MVEX_START;
		return (data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.FLAGS1_INDEX] & MvexInfoFlags1.NDS) != 0;
	}

	/** DO NOT USE: INTERNAL API */
	public static boolean canUseEvictionHint(int code) {
		int index = code - IcedConstants.MVEX_START;
		return (data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.FLAGS1_INDEX] & MvexInfoFlags1.EVICTION_HINT) != 0;
	}

	/** DO NOT USE: INTERNAL API */
	public static boolean canUseImmRoundingControl(int code) {
		int index = code - IcedConstants.MVEX_START;
		return (data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.FLAGS1_INDEX] & MvexInfoFlags1.IMM_ROUNDING_CONTROL) != 0;
	}

	/** DO NOT USE: INTERNAL API */
	public static boolean canUseRoundingControl(int code) {
		int index = code - IcedConstants.MVEX_START;
		return (data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.FLAGS1_INDEX] & MvexInfoFlags1.ROUNDING_CONTROL) != 0;
	}

	/** DO NOT USE: INTERNAL API */
	public static boolean canUseSuppressAllExceptions(int code) {
		int index = code - IcedConstants.MVEX_START;
		return (data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.FLAGS1_INDEX] & MvexInfoFlags1.SUPPRESS_ALL_EXCEPTIONS) != 0;
	}

	/** DO NOT USE: INTERNAL API */
	public static boolean getIgnoresOpMaskRegister(int code) {
		int index = code - IcedConstants.MVEX_START;
		return (data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.FLAGS1_INDEX] & MvexInfoFlags1.IGNORES_OP_MASK_REGISTER) != 0;
	}

	/** DO NOT USE: INTERNAL API */
	public static boolean getRequireOpMaskRegister(int code) {
		int index = code - IcedConstants.MVEX_START;
		return (data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.FLAGS1_INDEX] & MvexInfoFlags1.REQUIRE_OP_MASK_REGISTER) != 0;
	}

	/** DO NOT USE: INTERNAL API */
	public static boolean getNoSaeRc(int code) {
		int index = code - IcedConstants.MVEX_START;
		return (data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.FLAGS2_INDEX] & MvexInfoFlags2.NO_SAE_ROUNDING_CONTROL) != 0;
	}

	/** DO NOT USE: INTERNAL API */
	public static boolean isConvFn32(int code) {
		int index = code - IcedConstants.MVEX_START;
		return (data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.FLAGS2_INDEX] & MvexInfoFlags2.CONV_FN32) != 0;
	}

	/** DO NOT USE: INTERNAL API */
	public static boolean getIgnoresEvictionHint(int code) {
		int index = code - IcedConstants.MVEX_START;
		return (data[index * MvexInfoData.STRUCT_SIZE + MvexInfoData.FLAGS2_INDEX] & MvexInfoFlags2.IGNORES_EVICTION_HINT) != 0;
	}

	/** DO NOT USE: INTERNAL API */
	public static int getTupleType(int code, int sss) {
		return MvexTupleTypeLut.data[getTupleTypeLutKind(code) * 8 + sss];
	}
}
